### PR TITLE
Category merge branch

### PR DESCRIPTION
**Pull request by tavisca-rrai**
@achirametavisca @tavisca-smohite

### CODE REVIEW

File: Tests/OnlineRetailPortal.Tests/ProductMockTests.cs [hunk 1]
```diff
@@ -115,7 +115,7 @@ private AddProductStoreRequest GetRequest()
                 Description = "Green Bottle",
                 HeroImage = "example.com" ,
                 Price = new Contracts.Price {Money = new Contracts.Money(99.99, "INR"), IsNegotiable = false },
-                Category = Contracts.Category.Other,
+                Category = new Contracts.Category() { Name="Others"},
                 Images = new List<string>() { "ex.com" },
```
Comment (Tavisca-skhan) [inline Tests/OnlineRetailPortal.Tests/ProductMockTests.cs:118]: Others or Other

File: src/OnlineRetailPortal.Contracts/Contracts/ICategoryService.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OnlineRetailPortal.Contracts
+{
+    public interface ICategoryService
+    {
+        Task<GetCategoriesServiceResponse> GetCategoriesAsync();
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Contracts/Contracts/ICategoryService.cs]: Rename GetCategoriesServiceResponse to CategoriesServiceResponse

File: src/OnlineRetailPortal.Contracts/Models/GetCategoriesStoreResponse.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Collections.Generic;
+
+namespace OnlineRetailPortal.Contracts
+{
+    public class GetCategoriesStoreResponse
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Contracts/Models/GetCategoriesStoreResponse.cs]: check class name

File: src/OnlineRetailPortal.Core/Models/Category.cs [hunk 4]
```diff
@@ -1,19 +1,29 @@
...
+        {
+            this.Name = name;
+        }
+        public int Id { get; set; }
+        public string Name { get; }
+
+        public async static Task<List<Category>> GetCategoriesAsync(ICategoryStore categoryStore)
+        {
+            var response = await categoryStore.GetCategoriesAsync();
+            var CoreCategoriesResponse = response.ToCoreResponse();
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Core/Models/Category.cs]: follow naming convention.

File: src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs [hunk 5]
```diff
@@ -1,39 +1,45 @@
...
-                case Category.Property:
-                    return Contracts.Category.Property;
-            }
-            throw new NotSupportedException(category + " is not supported");
+            /*
+             switch (category)
+             {
+                 case Category.Bike:
+                     return Contracts.Category.Bike;
+                 case Category.Book:
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs]: remove commented code

File: src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs [hunk 6]
```diff
@@ -56,6 +62,9 @@ public static Category ToModel(this Contracts.Category category)
                     return Category.Property;
             }
             throw new NotSupportedException(category + " is not supported");
-        }
+            */
+            Category coreCategory = new Category(category.Name);
+            return coreCategory;
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs]: why translation of just name not id?

File: src/OnlineRetailPortal.Web/Controllers/CategoryController.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            try {
+
+              var  serviceResponse = await _categoryService.GetCategoriesAsync();
+                 return serviceResponse.ToCategoriesContract();
+            }
+            catch(Exception )
+            {
+                //Log(exception.Message, exception.trace);
+                throw new BaseException(500,"Internal Server Error", null,System.Net.HttpStatusCode.NotFound);
+
```
Comment (tavisca-rnayyar) [inline src/OnlineRetailPortal.Web/Controllers/CategoryController.cs]: Do we really need a catch here?

File: src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using System.Text;
+
+namespace OnlineRetailPortal.Core
+
+{
+    public static class GetCategoryCoreResponseTranslator
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs]: check name?

File: src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        public static List<Category> ToCoreResponse(this Contracts.GetCategoriesStoreResponse getCategoriesStoreResponse)
+        {
+            var categories = new List<Category>();
+            foreach(var x in getCategoriesStoreResponse.Categories)
+            {
+                categories.Add(new Category(x.Name) { Id = x.Id });
+            }
+            var response = categories;
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs]: what is the use of var response?

File: src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace OnlineRetailPortal.Web
+{
+    public static class ServiceResponseTranslator
+    {
+        public static CategoryResponse ToCategoriesContract(this Contracts.GetCategoriesServiceResponse getCategoriesServiceResponse)
+        {
+            CategoryResponse response = new CategoryResponse()
+            {
+                Categories = getCategoriesServiceResponse.Categories.Select(x => new Category()
```
Comment (tavisca-rnayyar) [inline src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs]: we might need a null check here for categories

File: src/OnlineRetailPortal.Web/Properties/launchSettings.json [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
-  "$schema": "http://json.schemastore.org/launchsettings.json",
+{
   "iisSettings": {
     "windowsAuthentication": false,
     "anonymousAuthentication": true,
     "iisExpress": {
       "applicationUrl": "http://localhost:62281",
-      "sslPort": 44357
+      "sslPort": 0
     }
```
Comment (Tavisca-skhan) [inline src/OnlineRetailPortal.Web/Properties/launchSettings.json]: why change in this file?

**Review by Tavisca-skhan — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs (hunk 10 above)
Comment (tavisca-rnayyar) [reply src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs]: And for getCategoriesServiceResponse as well

File: src/OnlineRetailPortal.Web/Translators/CategoryTranslator.cs [hunk 12]
```diff
@@ -29,11 +30,17 @@ public static Contracts.Category ToEntity(this string category)
...
             }
+
             throw new NotSupportedException(category + " is not supported");
+            */
+            Contracts.Category contractsCategory = new Contracts.Category();
+            contractsCategory.Name = category;
+            return contractsCategory;
         }
 
         public static string ToModel(this Contracts.Category category)
```
Comment (tavisca-rnayyar) [inline src/OnlineRetailPortal.Web/Translators/CategoryTranslator.cs:18]: Null checks are required in all the translators where required

File: src/OnlineRetailPortal.Services/Translators/GetCategoriesServiceResponseTranslator.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+using System.Linq;
+using System.Text;
+
+namespace OnlineRetailPortal.Services
+{
+    public static class GetCategoriesServiceResponseTranslator
+    {
+        public static GetCategoriesServiceResponse ToCategoriesResponse(this List<Core.Category> categories)
+        {
+            GetCategoriesServiceResponse response = new GetCategoriesServiceResponse()
```
Comment (tavisca-rnayyar) [inline src/OnlineRetailPortal.Services/Translators/GetCategoriesServiceResponseTranslator.cs]: Check if need NullCheck here

**Review by tavisca-rnayyar — CHANGES_REQUESTED** [review]
Please revise and add null checks where ever required.

File: Tests/OnlineRetailPortal.Tests/ProductMockTests.cs (hunk 1 above)
Comment (tavisca-rrai) [reply Tests/OnlineRetailPortal.Tests/ProductMockTests.cs:118]: Others

File: src/OnlineRetailPortal.Contracts/Contracts/ICategoryService.cs (hunk 2 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Contracts/Contracts/ICategoryService.cs]: should i also change GetCategoriesServiceResponseTranslator  to  CategoriesServiceResponseTranslator  in Translator folder

File: src/OnlineRetailPortal.Core/Models/Category.cs (hunk 4 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Models/Category.cs]: okay

File: src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs (hunk 8 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs]: should i change this to CategoryCoreResponseTranslator

File: src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs (hunk 9 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Translators/GetCategoryCoreResponseTranslator.cs]: just to increase readability

File: src/OnlineRetailPortal.Web/Properties/launchSettings.json (hunk 11 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Web/Properties/launchSettings.json]: shekhar did this , to implement cors

File: src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs (hunk 6 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs]: id will be generated from database,

File: src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs (hunk 5 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Translators/CategoryTranslator.cs]: removed

File: src/OnlineRetailPortal.Web/Controllers/CategoryController.cs (hunk 7 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Web/Controllers/CategoryController.cs]: removed

File: src/OnlineRetailPortal.Contracts/Models/GetProductResponse.cs [hunk 14]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlineRetailPortal.Contracts
```
Comment (achirametavisca) [inline src/OnlineRetailPortal.Contracts/Models/GetProductResponse.cs]: delete this file

File: src/OnlineRetailPortal.Core/Models/Category.cs [hunk 15]
```diff
@@ -1,19 +1,28 @@
...
+        {
+            this.Name = name;
+        }
+
+        public string Name { get; }
+
+        public async static Task<List<Category>> GetCategoriesAsync(ICategoryStore categoryStore)
+        {
+            var response = await categoryStore.GetCategoriesAsync();
+            var coreCategoriesResponse = response.ToCoreResponse();
```
Comment (achirametavisca) [inline src/OnlineRetailPortal.Core/Models/Category.cs]: ToModel

File: src/OnlineRetailPortal.Mock/Models/MockCategoryStore.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OnlineRetailPortal.Mock
+{
+    public class MockCategoryStore : ICategoryStore
+    {
+        
+        public List<Category> Categories;
```
Comment (achirametavisca) [inline src/OnlineRetailPortal.Mock/Models/MockCategoryStore.cs]: make it private

File: src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs [hunk 17]
```diff
@@ -9,6 +9,7 @@ public static class CategoryTranslator
     {
         public static string ToEntity(this Category category)
         {
+            /*
             switch (category)
```
Comment (achirametavisca) [inline src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs]: delete commented code

File: src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs [hunk 18]
```diff
@@ -31,12 +32,18 @@ public static string ToEntity(this Category category)
                     return "Property";
             }
             throw new NotSupportedException(category + " is not supported");
+            */
+            return category.Name;
         }
 
         public static Category ToModel(this string category)
         {
+            Category res = new Category() { Name = category };
```
Comment (achirametavisca) [inline src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs]: delete commented code

File: src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs (hunk 17 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs]: done

File: src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs (hunk 10 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Web/Translators/GetCategoriesWebResponse.cs]: done

File: src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs (hunk 18 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.MongoDBStore/Translator/CategoryTranslator.cs]: done

File: src/OnlineRetailPortal.Mock/Models/MockCategoryStore.cs (hunk 16 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Mock/Models/MockCategoryStore.cs]: done

File: src/OnlineRetailPortal.Core/Models/Category.cs (hunk 15 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Core/Models/Category.cs]: i think it should be ToEntity

File: src/OnlineRetailPortal.Contracts/Models/GetProductResponse.cs (hunk 14 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Contracts/Models/GetProductResponse.cs]: deleted

File: src/OnlineRetailPortal.Contracts/Models/GetCategoriesStoreResponse.cs (hunk 3 above)
Comment (tavisca-rrai) [reply src/OnlineRetailPortal.Contracts/Models/GetCategoriesStoreResponse.cs]: done

**Review by tavisca-swapnil — APPROVED** [review]

**Review by tavisca-mrathore — APPROVED** [review]
good to go.

File: Contracts/v1.0/get-categories/response.json5 [hunk 19]
```diff
@@ -0,0 +1,17 @@
+//Response Data
+{
+    "Categories" : [
+            "Accessory",
```
Comment (tavisca-acharpe) [inline Contracts/v1.0/get-categories/response.json5:4]: Send List of category object which contains ->1) category name 2) list of tags

File: src/OnlineRetailPortal.Web/Translators/CategoryResponseTranslator.cs [hunk 20]
```diff
@@ -9,13 +9,27 @@ public static class CategoryResponseTranslator
...
-            CategoryResponse response = new CategoryResponse() { Categories = new List<string>() { } };
-            
+            /*
+            CategoryResponse response = new CategoryResponse() { 
+                Categories = getCategoriesServiceResponse.Categories.Select(x => new Category()
+                {
+                    Name = x.Name,
+                    Tags = x.Tags
+                }).ToList(),
+            };
```
Comment (tavisca-acharpe) [inline src/OnlineRetailPortal.Web/Translators/CategoryResponseTranslator.cs]: Remove commented code

**Review by tavisca-acharpe — APPROVED** [review]

**Pull request merged**